### ISSUE TITLE
AbilityDamageEntityEvent called even when no damage inflicted

### ISSUE THREAD

**Issue by MKieburtz** [ProjectKorra/ProjectKorra#1162]
## Intro
I've noticed an issue regarding how PK calls its `AbilityDamageEntityEvent`. Specifically this event is being called multiple times even when the entity supposedly being damaged is not actually damaged! This is bad because this event should not ever be called when entities don't actually take damage from abilities.

## Evidence
Here's are some simple print outs for when the event is called. Simply printing out what entity got damaged, how much health it had _before the damage is inflicted_, and the current server tick.
```Java
    @EventHandler
    public void onAbilityDamageEntityEvent(AbilityDamageEntityEvent event) {
        System.out.println(event.getSource().getName() + " Damaged " + event.getEntity().getName()
                + " with " + event.getAbility().getName() + " for " + event.getDamage());
        LivingEntity livingEntity = (LivingEntity) event.getEntity();
        System.out.println("remaining health on entity: " + livingEntity.getHealth());
        System.out.println("current tick: " + Bukkit.getServer().getCurrentTick());
    }
```
Here's what happens when I shoot an _uncharged_ Air Swipe at a zombie just once:
```
[19:27:53] [Server thread/INFO]: BitNinjaxD Damaged Zombie with AirSwipe for 2.0
[19:27:53] [Server thread/INFO]: remaining health on entity: 20.0
[19:27:53] [Server thread/INFO]: current tick: 809
[19:27:53] [Server thread/INFO]: BitNinjaxD Damaged Zombie with AirSwipe for 2.0
[19:27:53] [Server thread/INFO]: remaining health on entity: 18.079999923706055
[19:27:53] [Server thread/INFO]: current tick: 810
[19:27:53] [Server thread/INFO]: BitNinjaxD Damaged Zombie with AirSwipe for 2.0
[19:27:53] [Server thread/INFO]: remaining health on entity: 18.079999923706055
[19:27:53] [Server thread/INFO]: current tick: 811
[19:27:53] [Server thread/INFO]: BitNinjaxD Damaged Zombie with AirSwipe for 2.0
[19:27:53] [Server thread/INFO]: remaining health on entity: 18.079999923706055
[19:27:53] [Server thread/INFO]: current tick: 812
[19:27:53] [Server thread/INFO]: BitNinjaxD Damaged Zombie with AirSwipe for 2.0
[19:27:53] [Server thread/INFO]: remaining health on entity: 18.079999923706055
[19:27:53] [Server thread/INFO]: current tick: 813
```
As you can see the event is called 5(!) times even though the zombie only took damage once. There were no other mobs near the zombie.

This does not happen with `EntityDamageByEntityEvent`, which is also called when an entity takes bending damage. Using this code to test:
```Java
    @EventHandler
    public void onEntityDamage(EntityDamageByEntityEvent event) {
        Entity entity = event.getEntity();
        System.out.println(event.getDamager().getName() + " Damaged " + event.getEntity().getName() + " for " + 
                                      event.getDamage() + " damage.");
        LivingEntity livingEntity = (LivingEntity)entity;
        System.out.println("remaining health on entity: " + livingEntity.getHealth());
        System.out.println("current tick: " + Bukkit.getServer().getCurrentTick());
    } 
``` 

Results in this when hit by Air Swipe 3 times:
```
[19:37:12 INFO]: BitNinjaxD Damaged Skeleton for 2.0 damage.
[19:37:12 INFO]: remaining health on entity: 17.61680030822754
[19:37:12 INFO]: current tick: 1073
[19:37:27 INFO]: BitNinjaxD Damaged Skeleton for 2.0 damage.
[19:37:27 INFO]: remaining health on entity: 15.616800308227539
[19:37:27 INFO]: current tick: 1363
[19:37:29 INFO]: BitNinjaxD Damaged Skeleton for 2.0 damage.
[19:37:29 INFO]: remaining health on entity: 13.616800308227539
[19:37:29 INFO]: current tick: 1404
```
This is what I would expect, just one damage event each time the entity is actually damaged. 
## Conclusion
This is a major bug. The `AbilityDamageEntityEvent` simply does not work at this time because it is unreliable. I believe this is worthy of a hotfix since every addon that relies on this event is effected by this issue (and may not even be aware of it).

## Final note
Air Swipe curiously calls the `DamageHandler::damageEntity` method inside of a `BukkitRunnable`, which may explain some of the behavior here. However the Lava Disk ability of Hyperion by @PrimordialMoros, does not do this, but still results in the same anomalous behavior. There is something else going on in my opinion.

Thanks for reading.

**Comment by Varhagna** [comment]
Hi, thank you for bringing this to our attention.

At an initial glance this is most likely due to how AirSwipe actually applies damage to entities. 

Since an AirSwipe is actually multiple streams which deal damage independently, we need to verify that we're not applying damage multiple times to the same entity (since its possible that an entity's location is within the collision radius of multiple streams). 

This is why it you see it called multiple times, once per stream which collided with the entity. 

All of that being said I agree this is something that should be fixed.

**Comment by MKieburtz** [comment]
Tomorrow I will investigate what other moves this happens with (I believe Torrent may be one). It'd be nice if there was a check implemented that would not call the event if it was used erroneously. I feel like it will be hard to get a bunch of addon developers to update their moves due to a bug that most players won't notice...

**Comment by Simplicitee** [comment]
Make sure to note whether the abilities you test intentionally (at least code-wise) deal damage multiple times

**Comment by MKieburtz** [comment]
The part of torrent where the water orbits around you is confirmed.

```
[19:21:23 INFO]: BitNinjaxD Damaged Skeleton with Torrent for 1.5
[19:21:23 INFO]: remaining health on entity: 12.5
[19:21:23 INFO]: current tick: 7027
[19:21:23 INFO]: BitNinjaxD Damaged Skeleton with Torrent for 1.5
[19:21:23 INFO]: remaining health on entity: 11.0
[19:21:23 INFO]: current tick: 7028
[19:21:23 INFO]: BitNinjaxD Damaged Skeleton with Torrent for 1.5
[19:21:23 INFO]: remaining health on entity: 11.0
[19:21:23 INFO]: current tick: 7028
[19:21:23 INFO]: BitNinjaxD Damaged Skeleton with Torrent for 1.5
[19:21:23 INFO]: remaining health on entity: 11.0
[19:21:23 INFO]: current tick: 7029
[19:21:23 INFO]: BitNinjaxD Damaged Skeleton with Torrent for 1.5
[19:21:23 INFO]: remaining health on entity: 11.0
[19:21:23 INFO]: current tick: 7029
```
This is what happens when the skeleton takes damage just once. It's interesting that the event fires twice during the same tick both times after the initial (correct) activation. This didn't happen with air swipe.

**Comment by MKieburtz** [comment]
I've found 2 addon moves so far that this happens with as well. And I'll also throw this one out there, here's arc spark from project addons, so I'll ping you here @Simplicitee. I'll use a pastebin for this one because it's long: https://pastebin.com/QLYnWgwE

So here the zombie gets damaged 5 times. But the event is called a whopping 29 times. Yikes.

**Comment by MKieburtz** [comment]
Just tried Project Addons flame breath as well, with a similar result: https://pastebin.com/vKf4YEYV.

6 actual damage events, but 85 calls to the event.

**Comment by Simplicitee** [comment]
As a guess, I'd say that this is happening because things try to use `damageEntity` during an entity's no damage ticks

**Comment by MKieburtz** [comment]
> As a guess, I'd say that this is happening because things try to use `damageEntity` during an entity's no damage ticks

Could the fix be as simple as just checking to see if the entity is in those no damage ticks before actually calling the event?

**Comment by Simplicitee** [comment]
> Could the fix be as simple as just checking to see if the entity is in those no damage ticks before actually calling the event?

If that is actually the problem AND if that's possible to check then yeah, I'd say it's a pretty good attempt

**Comment by MKieburtz** [comment]
https://hub.spigotmc.org/javadocs/spigot/org/bukkit/entity/LivingEntity.html#getNoDamageTicks()

Seems promising.

**Comment by ChristopherWMM** [comment]
Fixed as of #1163
